internal/hcl/evaluator.go

```diff
@@ -406,33 +406,78 @@ func (e *Evaluator) expandBlocks(blocks Blocks, lastContext hcl.EvalContext) Blo
 }
 
 func (e *Evaluator) expandDynamicBlocks(blocks ...*Block) Blocks {
+	var newBlocks Blocks
+
 	for _, b := range blocks {
-		e.expandDynamicBlock(b)
+		if b.HasDynamicBlock() {
+			newBlocks = append(newBlocks, e.expandDynamicBlock(b))
+			continue
+		}
+
+		newBlocks = append(newBlocks, b)
 	}
-	return blocks
+
+	return newBlocks
 }
 
-func (e *Evaluator) expandDynamicBlock(b *Block) {
-	for _, sub := range b.Children() {
-		e.expandDynamicBlock(sub)
-	}
+func (e *Evaluator) expandDynamicBlock(b *Block) *Block {
+	var newChildBlocks Blocks
+	for _, child := range b.Children() {
+		// handle nested dynamic blocks
+		child = e.expandDynamicBlock(child)
 
-	for _, sub := range b.Children().OfType("dynamic") {
-		e.logger.Debug().Msgf("expanding block %s because a dynamic block was found %s", b.LocalName(), sub.LocalName())
+		// if the child block is not a dynamic there is nothing to do
+		// so add it to the newBlocks straight away and continue.
+		if child.Type() != "dynamic" {
+			newChildBlocks = append(newChildBlocks, child)
+			continue
+		}
 
-		blockName := sub.TypeLabel()
-		// Remove all the child blocks with the blockName so that we don't inject the expanded blocks twice.
-		// This could happen if a module "reloads" and the input variables change.
-		b.RemoveDynamicBlocks(blockName)
+		e.logger.Debug().Msgf("expanding block %s because a dynamic block was found %s", b.LocalName(), child.LocalName())
+		blockName := child.TypeLabel()
 
-		expanded := e.expandBlockForEaches([]*Block{sub})
+		// for each expanded dynamic block add the generated "content" as a
+		// new child block into the parent block.
+		expanded := e.expandBlockForEaches([]*Block{child})
 		for _, ex := range expanded {
-			if content := ex.GetChildBlock("content"); content != nil {
-				_ = e.expandDynamicBlocks(content)
-				b.InjectBlock(content, blockName)
+			content := ex.GetChildBlock("content")
+			if content == nil {
+				continue
+			}
+
+			_ = e.expandDynamicBlocks(content)
+
+			content.SetLabels([]string{})
+			content.SetType(blockName)
+
+			for attrName, attr := range content.AttributesAsMap() {
+				b.context.Root().SetByDot(attr.Value(), fmt.Sprintf("%s.%s.%s", b.Reference().String(), blockName, attrName))
 			}
+
+			newChildBlocks = append(newChildBlocks, content)
 		}
 	}
+
+	return &Block{
```

Comment:
return a new pointer to a block here as the old `Evaluator` logic "reloads" modules using a list of original `RawBlocks`. If we modify the original block passed into this method (which is a pointer) it alters both the `Blocks` and `RawBlocks` lists and causes issues when we run the module again as the child `Block` doesn't have the correct context.